p2p/chain.py

```diff
@@ -0,0 +1,242 @@
+        # find the common ancestor between our chain and the peer's.
+        start_at = max(0, head.block_number - eth.MAX_HEADERS_FETCH)
+        self.logger.debug("Asking %s for header batch starting at %d", peer, start_at)
+        peer.sub_proto.send_get_block_headers(start_at, eth.MAX_HEADERS_FETCH, reverse=False)
+        max_consecutive_timeouts = 3
+        consecutive_timeouts = 0
+        while True:
+            try:
+                headers = await wait_with_token(
+                    [self._new_headers.get(), peer.wait_until_finished()],
+                    self.cancel_token,
+                    timeout=3)
+            except OperationCancelled:
+                break
+            except TimeoutError:
+                self.logger.debug("Timeout waiting for header batch from %s", peer)
+                consecutive_timeouts += 1
+                if consecutive_timeouts > max_consecutive_timeouts:
+                    self.logger.debug(
+                        "Too many consecutive timeouts waiting for header batch, aborting sync "
+                        "with %s", peer)
+                    break
+                continue
+
+            if peer.is_finished():
+                self.logger.debug("%s disconnected, stopping sync", peer)
+                break
+
+            consecutive_timeouts = 0
+            if headers[-1].block_number <= start_at:
+                self.logger.debug(
+                    "Ignoring headers from %d to %d as they've been processed already",
+                    headers[0].block_number, headers[-1].block_number)
+                continue
+
+            # TODO: Process headers for consistency.
+            # TODO: Queue body/receipt downloads.
+            for header in headers:
+                await self.chaindb.coro_persist_header_to_db(header)
+                start_at = header.block_number
+
+            self.logger.debug("Asking %s for header batch starting at %d", peer, start_at)
+            # TODO: Instead of requesting sequential batches from a single peer, request a header
+            # skeleton and make concurrent requests, using as many peers as possible, to fill the
+            # skeleton.
+            peer.sub_proto.send_get_block_headers(start_at, eth.MAX_HEADERS_FETCH, reverse=False)
+
+    async def wait_until_finished(self) -> None:
+        start_at = time.time()
+        # Wait at most 5 seconds for pending peers to finish.
+        self.logger.info("Waiting for %d running peers to finish", len(self._running_peers))
+        while time.time() < start_at + 5:
+            if not self._running_peers:
+                break
+            await asyncio.sleep(0.1)
+        else:
+            self.logger.info("Waited too long for peers to finish, exiting anyway")
+
+    async def stop(self) -> None:
+        self.cancel_token.trigger()
+        self.peer_pool.unsubscribe(self)
+        await self.wait_until_finished()
+
+    def handle_msg(self, peer: ETHPeer, cmd: protocol.Command,
```

Comment:
I just learned about [`functools.single_dispatch`](https://docs.python.org/3/library/functools.html#functools.singledispatch) which would in *theory* allow you to split the handlers here up into many different single functions.

I'm not sure this is the right use case, and I suspect the fact it's a class method might make it a slightly more complex use case.

Reply:
Or maybe this library if you want multiple dispatch.  http://multiple-dispatch.readthedocs.io/en/latest/index.html though I'm not keen on adding the dependency